Remove delimiter-less ToDelimitedString overloads

atifaziz opened the pull request:
This PR addresses #229.

The overloads have been obsolete since [version 2.1](https://github.com/morelinq/MoreLINQ/releases/tag/v2.1.0).

fsateler commented on MoreLinq/ToDelimitedString.cs:
```diff
@@ -25,37 +25,14 @@ namespace MoreLinq
         /// simple ToString() conversion.</param>
-        /// <param name="delimiter">The delimiter to inject between elements. May be null, in which case
-        /// the executing thread's current culture's list separator is used.</param>
+        /// <param name="delimiter">The delimiter to inject between elements.</param>
```
The new `<exception>` tag for the delimiter should be added.

atifaziz replied:
If you insist :) although we haven't been doing that consistently.

fsateler commented:
> Looks good, except that the new exception should be documented, and these breaking changes should be documented somewhere as an upgrade guide.

Should we add a `CHANGES.md` file ?

atifaziz commented:
> Should we add a `CHANGES.md` file ?

Unless someone can help automate that through a discipline & process (as I've seen employed in some projects) we use then I don't want to add more to my plate in terms of maintenance at this point. Having said, I'm currently relying on [milestones](https://github.com/morelinq/MoreLINQ/milestones) to help. Each [release](https://github.com/morelinq/MoreLINQ/releases) is linked to its milestone. [Issues that introduce breaking changes are marked so](https://github.com/morelinq/MoreLINQ/issues?q=is%3Aissue+is%3Aopen+label%3Abreaking-change). So for release 3.0, I'm expecting that people will be able to quickly consult [the breaking changes there](https://github.com/morelinq/MoreLINQ/issues?utf8=%E2%9C%93&q=is%3Aissue+milestone%3A3.0.0+label%3Abreaking-change+) and follow-up the rational that's been captured in the discussion of the issue that interests them. I admit this doesn't answer the “upgrade guide” bit but it's better than nothing. And as I said, if I can get another pair of hands then I'd vote for a `CHANGES.md` too. However, I feel that #284 (contribution guide) and #247 (style guide) might trump that in priority. What do you think?

fsateler commented:
> Unless someone can help automate that through a discipline & process (as I've seen employed in some projects) we use then I don't want to add more to my plate in terms of maintenance at this point. 

I can totally understand this.

> Having said, I'm currently relying on milestones to help. Each release is linked to its milestone. Issues that introduce breaking changes are marked so. So for release 3.0, I'm expecting that people will be able to quickly consult the breaking changes there and follow-up the rational that's been captured in the discussion of the issue that interests them. I admit this doesn't answer the “upgrade guide” bit but it's better than nothing. And as I said, if I can get another pair of hands then I'd vote for a CHANGES.md too. However, I feel that #284 (contribution guide) and #247 (style guide) might trump that in priority. What do you think?

I don't think the CHANGES.md file needs to be too high maintainance. I think simply reusing the content put in the github release page would be sufficient. This has the advantage that once this file is in existence, it can be filled by the PR submitter, thus overall reducing the release load (as then there's only a copy-paste into the github release description).

I think I used the wrong description with "upgrade guide". A thourough guide is not what I meant. I mean a simple bullet-point list of things to take care when upgrading. In this case, the entry could be as simple as:

- The `ToDelimitedString` overload with no delimiter has been removed. If you want to preserve the exact behavior as previously, pass `TextInfo.ListSeparator` as delimiter.
- `ToDelimitedString` no longer accepts `null` as its delimiter parameter. Pass `TextInfo.ListSeparator` as delimiter if you want to mimic the old behavior.

Does this sound reasonably low-overhead?

atifaziz commented:
> This has the advantage that once this file is in existence, it can be filled by the PR submitter

Right but this create a process that needs to be documented in a contribution guide. The advantage of release pages is that they can be changed, updated & improved _after_ the release whereas a `CHANGES.md` that's part of the repo kind of pollutes the project history and would need maintenance of release branches. Perhaps a compromise would be a wiki per release that can be community edited? In any case, I feel we should be having this discussion outside of this PR. Speaking of the PR, we good to merge this? I think I've addressed everything from [your review](https://github.com/morelinq/MoreLINQ/pull/373#pullrequestreview-93364600).

fsateler commented:
> Right but this create a process that needs to be documented in a contribution guide. 

Right.

> The advantage of release pages is that they can be changed, updated & improved after the release whereas a CHANGES.md that's part of the repo kind of pollutes the project history and would need maintenance of release branches. Perhaps a compromise would be a wiki per release that can be community edited? 

This might be a good idea.

> In any case, I feel we should be having this discussion outside of this PR. Speaking of the PR, we good to merge this? I think I've addressed everything from your review.

Yes, LGTM.
